### PR TITLE
Setup for Jax dependency

### PR DESCRIPTION
**Pull request by VIGNESHinZONE**
Creating this based on the reviews from https://github.com/deepchem/deepchem/pull/2549

I have added `deepchem` Jax support. Installs can be made either using

```
pip install -e `.[jax]`
```
or
```
pip install 'deepchem[jax]'
```
This is just a temporary solution - 
- `jax_setup.yml` has been copied from `main.yml` 
- `env_jax.yml` also has `TensorFlow` dependency because `deepchem` is dependent on `TensorFlow` and I cant run `import deepchem` without tensorflow installed.

### CODE REVIEW

**Comment by VIGNESHinZONE** [comment]
@rbharath , Could you please take a look at this - 

1. All of deepchem's unit test is written using `unittest` library and called using `pytest` in CI
2. I have added a test for checking `jax` imports using `pytest` markers with tag `pyest.mark.jax`
3. For some reason, all the tests including the `Unittest` are being called

**Comment by peastman** [comment]
> env_jax.yml also has TensorFlow dependency because deepchem is dependent on TensorFlow and I cant run import deepchem without tensorflow installed.

We should fix that.  I opened issue #2561 to track it.

**Comment by ncfrey** [comment]
> @rbharath @ncfrey , Could you please take a look at this -
> 
> All of deepchem's unit test is written using `unittest` library and called using `pytest` in CI. I have added a test for checking `jax` imports using `pytest` markers with tag `pyest.mark.jax`. For some reason, all the tests including the `Unittest` are being called despite me using the correct commands as per the `pytest` docs https://docs.pytest.org/en/6.2.x/example/markers.html

PR looks good! For this, when you do `pytest --markers` locally do you see the `jax` marker?

File: .github/workflows/jax_setup.yml
Comment (rbharath) [inline .github/workflows/jax_setup.yml:32]: I think this line is outdated (in the original as well). Can you swap this to `pip install tensorflow==2.4` for now?

File: env_jax.yml
Comment (rbharath) [inline env_jax.yml:7]: Does the latest jax run on numpy 1.20 or 1.19? You can test this by installing jax in a fresh environment and seeing which numpy it pulls down

**Review by rbharath — COMMENTED** [review]
This is looking good! A couple of minor comments here

File: env_jax.yml
Comment (VIGNESHinZONE) [reply env_jax.yml:7]: Yup, it works well.

**Comment by VIGNESHinZONE** [comment]
@ncfrey @rbharath 
Thanks for the review, i had missed adding the `jax` marker in `setup.cfg` file. `Pytest` is still failing but this time, it fails because its including the `torch` tests and has deselected 729 tests.

File: env_jax.yml
Comment (rbharath) [reply env_jax.yml:7]: Do you mean the latest jax uses numpy 1.19?

**Review by rbharath — COMMENTED** [review]
Perhaps we need to manually add pytorch markers to the pytorch pytests to fix?

File: env_jax.yml
Comment (VIGNESHinZONE) [reply env_jax.yml:7]: I'm sorry if I might have misunderstood the question, but from its `setup.py` file it doesn't mention any specific depedency on numpy 1.20 or 1.19 - https://github.com/google/jax/blob/6382cee7084ebea2628ca139cbaf7415108cd78f/setup.py#L37

I even ran a simple collab notebook and it works perfectly fine in `numpy 1.19`  with latest version of `jax 2.14`- [google collab](https://colab.research.google.com/drive/1Iwd2B8wTqLBQSS4t4OWvmh14p5jSCAD_?usp=sharing)

**Comment by VIGNESHinZONE** [comment]
@rbharath , All Jax test cases are passing. Please let me know if we can move ahead with the merging it.
The problem was with imports in few test for optional dependencies like `torch`, `transformer`, `xgboost`

Here is an example of optional depedency not being `import`'ed in `try` & `catch` blocks- 
https://github.com/deepchem/deepchem/blob/cbab9678fa8aa480996bcb030f7fedcd9950ca3d/deepchem/models/tests/test_gcn.py#L8-L17

But here is the `gcn.py` file - 

https://github.com/deepchem/deepchem/blob/cbab9678fa8aa480996bcb030f7fedcd9950ca3d/deepchem/models/torch_models/gcn.py#L4-L5

`GCNModel` also needs to be imported in the try catch blocks because it has `torch` imports.

**Review by rbharath — APPROVED** [review]
Looks good! Will hold off a bit in case @ncfrey @vsomnath have review comments otherwise we can merge this by this afternoon/evening (PST)

**Comment by VIGNESHinZONE** [comment]
Sorry for the extra commits after approval, had to merge with the `master` branch.  All `deepchem[jax]` test checks are passing

**Comment by VIGNESHinZONE** [comment]
`Deepchem[jax]` is having installation faults in windows, but it a problem with a library itself mentioned - https://github.com/google/jax/issues/6637#issuecomment-865386555

I still think it can be merged, because its running well on `Ubuntu 3.6 , 3.7` and `JaxModel` is still under development.

**Comment by rbharath** [comment]
Ok going ahead and merging! We can fix any issues in downstream PRs

**Comment by peastman** [comment]
I don't think this was ready to merge.  The new CI builds it added are failing.  The windows-latest build fails at the installation stage with

```
ERROR: Could not find a version that satisfies the requirement jaxlib==0.1.67 (from deepchem==2.6.0.dev20210622190958) (from versions: none)
ERROR: No matching distribution found for jaxlib==0.1.67 (from deepchem==2.6.0.dev20210622190958)
```

**Comment by VIGNESHinZONE** [comment]
I'm sorry for continuing with this PR. 
I had assumed that this breakage would only affect `jax windows` environment and not others as it's completely independent. Also, even `jax` team has acknowledged that it has faults with `windows build` https://github.com/google/jax/issues/6637#issuecomment-865386555

**Comment by peastman** [comment]
If it isn't expected to work on Windows, should we remove `windows-latest` from the build matrix?

https://github.com/deepchem/deepchem/blob/eab6ee63c1b3e958abe72a40efb2f5064201731b/.github/workflows/jax_setup.yml#L42

**Comment by VIGNESHinZONE** [comment]
yeah sure we could do that for now and monitor its compatibility in `windows` in the next release of `Jax`

**Pull request merged**
